File: src/tools/bash.go
package tools

import (
	"bytes"
	"errors"
	"os/exec"
	"time"
)

var BashTool = Tool{
	Type: "function",
	Function: ToolFunction{
		Name: "bash_tool",
		Description: `Execute shell commands in the workspace environment.  
Used for inspecting the filesystem, building projects, running tests, or executing scripts.`,
		Parameters: JSONSchema{
			Type: "object",
			Properties: map[string]Schema{
				"message": {
					Type:        "string",
					Description: "A message describing the command being executed by the llm",
				},
				"command": {
					Type:        "string",
					Description: "the command to be executed",
				},
				"working_directory": {
					Type:        "string",
					Description: "The current working directory",
				},
				"timeout_seconds": {
					Description: "seconds after which the command will be terminated automatically",
					Type:        "integer",
				},
			},
			Required: []string{"message", "command"},
		},
	},
}

// ============================
// BASH TOOL
// ============================

// Input

type BashInput struct {
	Message          string `json:"message"`
	Command          string `json:"command"`
	WorkingDirectory string `json:"working_directory,omitempty"`
	TimeoutSeconds   int    `json:"timeout_seconds,omitempty"`
}

// Output

type BashOutput struct {
	ExitCode   int    `json:"exit_code"`
	Stdout     string `json:"stdout"`
	Stderr     string `json:"stderr"`
	DurationMs int    `json:"duration_ms"`
}

// execution function
func RunBash(input BashInput) (BashOutput, error) {

	if input.Command == "" {
		return BashOutput{}, errors.New("command cannot be empty")
	}

	timeout := input.TimeoutSeconds

	if timeout == 0 {
		timeout = 30
	}

	start := time.Now()

	cmd := exec.Command("bash", "-c", input.Command)

	if input.WorkingDirectory != "" {
		cmd.Dir = input.WorkingDirectory
	}

	var stdout bytes.Buffer
	var stderr bytes.Buffer

	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()

	exitCode := 0

	if err != nil {
		if exitError, ok := err.(*exec.ExitError); ok {
			exitCode = exitError.ExitCode()
		} else {
			return BashOutput{}, err
		}
	}

	duration := time.Since(start)

	return BashOutput{
		ExitCode:   exitCode,
		Stdout:     stdout.String(),
		Stderr:     stderr.String(),
		DurationMs: int(duration.Milliseconds()),
	}, nil

}

File: src/tools/bash/bash.py
import json
import subprocess
import argparse


def run() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--message", type=str, required=True)
    parser.add_argument("--command", type=str, required=True)
    parser.add_argument("--working_directory", type=str, default=".")
    parser.add_argument("--timeout_seconds", type=int, default=10)

    args = parser.parse_args()

    print(args.command.replace('"',"\""))
    try:
        result = subprocess.run(
            ["bash", "-lc", args.command.replace('"',"\"")],
            cwd=args.working_directory,
            capture_output=True,
            text=True,
            timeout=args.timeout_seconds
        )

        print(json.dumps({
            "ok": True,
            "exit_code": result.returncode,
            "stdout": result.stdout,
            "stderr": result.stderr
        }))

    except subprocess.TimeoutExpired as e:
        print(json.dumps({
            "ok": False,
            "error": "Command timed out",
            "stdout": e.stdout or "",
            "stderr": e.stderr or ""
        }))

    except Exception as e:
        print(json.dumps({
            "ok": False,
            "error": str(e)
        }))


run()